### PR TITLE
More linting and formatting

### LINKED ISSUE #2069: Flake8 issues not covered so far
These are the `flake8` warnings that are ignored by `ruff` so far:

```
$ flake8 --max-line-length 120 --max-doc-length 100 --ignore=E221,E226,E241,E242,E402,E731,E741,F403,F405,F722,W503,D10,D40,D414 . | awk '{print $2}' | sort | uniq
E127
E128
E227
E231
E261
E301
E302
E303
E305
E722
F401
W391
```

Here are also the error messages:

```
$ flake8 --max-line-length 120 --max-doc-length 100 --ignore=E221,E226,E241,E242,E402,E731,E741,F403,F405,F722,W503,D10,D40,D414 .
./docs/source/k3d_sphinx.py:49:5: E303 too many blank lines (2)
./src/pymortests/base.py:66:41: E127 continuation line over-indented for visual indent
./src/pymortests/base.py:84:1: E302 expected 2 blank lines, found 1
./src/pymortests/__init__.py:4:1: W391 blank line at end of file
./src/pymortests/reductors/__init__.py:4:1: W391 blank line at end of file
./src/pymortests/parameters/__init__.py:4:1: W391 blank line at end of file
./src/pymortests/fixtures/__init__.py:4:1: W391 blank line at end of file
./src/pymortests/fixtures/model.py:40:24: E127 continuation line over-indented for visual indent
./src/pymortests/fixtures/operator.py:606:1: E302 expected 2 blank lines, found 1
./src/pymortests/fixtures/operator.py:632:1: E302 expected 2 blank lines, found 1
./src/pymortests/fixtures/operator.py:656:1: E302 expected 2 blank lines, found 1
./src/pymortests/discretizers/__init__.py:4:1: W391 blank line at end of file
./src/pymortests/algorithms/__init__.py:4:1: W391 blank line at end of file
./src/pymortests/core/__init__.py:4:1: W391 blank line at end of file
./src/pymortests/core/config.py:14:1: E303 too many blank lines (3)
./src/pymortests/analyticalproblems/__init__.py:4:1: W391 blank line at end of file
./src/pymortests/operators/__init__.py:4:1: W391 blank line at end of file
./src/pymortests/bindings/__init__.py:4:1: W391 blank line at end of file
./src/pymortests/models/__init__.py:4:1: W391 blank line at end of file
./src/pymor/__init__.py:49:1: E305 expected 2 blank lines after class or function definition, found 1
./src/pymor/__init__.py:51:1: F401 'pymor.core.config.config' imported but unused
./src/pymor/__init__.py:84:9: E301 expected 1 blank line, found 0
./src/pymor/reductors/__init__.py:4:1: W391 blank line at end of file
./src/pymor/reductors/neural_network.py:358:55: E127 continuation line over-indented for visual indent
./src/pymor/reductors/neural_network.py:362:55: E127 continuation line over-indented for visual indent
./src/pymor/reductors/neural_network.py:1316:43: E127 continuation line over-indented for visual indent
./src/pymor/parameters/__init__.py:4:1: W391 blank line at end of file
./src/pymor/discretizers/__init__.py:4:1: W391 blank line at end of file
./src/pymor/discretizers/fenics/__init__.py:5:1: F401 'pymor.discretizers.fenics.cg.discretize_stationary_cg' imported but unused
./src/pymor/discretizers/fenics/__init__.py:6:1: F401 'pymor.discretizers.fenics.domaindiscretizer.discretize_domain' imported but unused
./src/pymor/discretizers/builtin/__init__.py:5:1: F401 'pymor.discretizers.builtin.cg.discretize_instationary_cg' imported but unused
./src/pymor/discretizers/builtin/__init__.py:5:1: F401 'pymor.discretizers.builtin.cg.discretize_stationary_cg' imported but unused
./src/pymor/discretizers/builtin/__init__.py:6:1: F401 'pymor.discretizers.builtin.fv.discretize_instationary_fv' imported but unused
./src/pymor/discretizers/builtin/__init__.py:6:1: F401 'pymor.discretizers.builtin.fv.discretize_stationary_fv' imported but unused
./src/pymor/discretizers/builtin/__init__.py:7:1: F401 'pymor.discretizers.builtin.grids.gmsh.load_gmsh' imported but unused
./src/pymor/discretizers/builtin/__init__.py:8:1: F401 'pymor.discretizers.builtin.grids.oned.OnedGrid' imported but unused
./src/pymor/discretizers/builtin/__init__.py:9:1: F401 'pymor.discretizers.builtin.grids.rect.RectGrid' imported but unused
./src/pymor/discretizers/builtin/__init__.py:10:1: F401 'pymor.discretizers.builtin.grids.tria.TriaGrid' imported but unused
./src/pymor/discretizers/builtin/cg.py:994:34: E261 at least two spaces before inline comment
./src/pymor/discretizers/builtin/domaindiscretizers/__init__.py:4:1: W391 blank line at end of file
./src/pymor/discretizers/builtin/domaindiscretizers/gmsh.py:67:27: E128 continuation line under-indented for visual indent
./src/pymor/discretizers/builtin/gui/__init__.py:4:1: W391 blank line at end of file
./src/pymor/discretizers/builtin/gui/jupyter/__init__.py:84:5: E722 do not use bare 'except'
./src/pymor/discretizers/builtin/grids/__init__.py:4:1: W391 blank line at end of file
./src/pymor/scripts/__init__.py:4:1: W391 blank line at end of file
./src/pymor/algorithms/__init__.py:4:1: W391 blank line at end of file
./src/pymor/core/base.py:234:27: E261 at least two spaces before inline comment
./src/pymor/core/pickle.py:93:21: E227 missing whitespace around bitwise or shift operator
./src/pymor/analyticalproblems/__init__.py:4:1: W391 blank line at end of file
./src/pymor/vectorarrays/list.py:438:31: E231 missing whitespace after ','
./src/pymor/operators/__init__.py:4:1: W391 blank line at end of file
./src/pymor/parallel/__init__.py:4:1: W391 blank line at end of file
./src/pymor/models/__init__.py:4:1: W391 blank line at end of file
./src/pymor/tools/io/__init__.py:10:1: F401 'pymor.tools.deprecated.Deprecated' imported but unused
./src/pymor/tools/io/__init__.py:12:1: F401 '.matrices.load_matrix' imported but unused
./src/pymor/tools/io/__init__.py:12:1: F401 '.matrices.save_matrix' imported but unused
./src/pymordemos/__init__.py:4:1: W391 blank line at end of file
```

### PR DESCRIPTION
**Pull request by pmli**
Closes #2069.

### CODE REVIEW

File: src/pymor/algorithms/adaptivegreedy.py [hunk 1]
```diff
@@ -337,7 +337,8 @@ def map_vertex_to_mu(self, vertex):
     def visualize(self, vertex_data=None, vertex_inds=None, center_data=None, center_inds=None, volume_data=None,
                   vertex_size=80, vmin=None, vmax=None, new_figure=True):
         if self.dim not in (2, 3):
-            raise ValueError('Can only visualize samples of dimension 2, 3')
+            msg = 'Can only visualize samples of dimension 2, 3'
+            raise ValueError(msg)
```
Comment (sdrave) [inline src/pymor/algorithms/adaptivegreedy.py:341]: Disagree.

File: src/pymor/algorithms/basic.py [hunk 2]
```diff
@@ -5,7 +5,7 @@
 """Module containing some basic but generic linear algebra algorithms."""
 
 import numpy as np
-import scipy
+import scipy.linalg as spla
```
Comment (sdrave) [inline src/pymor/algorithms/basic.py:8]: Disagree.

File: src/pymor/algorithms/basic.py [hunk 3]
```diff
@@ -47,9 +47,8 @@ def almost_equal(U, V, product=None, sup_norm=False, rtol=1e-14, atol=1e-14):
     X = V.copy()
 
     # broadcast if necessary
-    if len(X) == 1:
-        if len(U) > 1:
-            X.append(X[np.zeros(len(U) - 1, dtype=int)])
+    if len(X) == 1 and len(U) > 1:
+        X.append(X[np.zeros(len(U) - 1, dtype=int)])
```
Comment (sdrave) [inline src/pymor/algorithms/basic.py:51]: While I prefer this, actually, I don't think the former should be an error.

File: src/pymor/algorithms/error.py [hunk 4]
```diff
@@ -206,9 +206,9 @@ def custom_value(rom, fom, reductor, mu, dim):
         result['rel_errors'] = rel_errors = errors / norms[:, :, np.newaxis]
         result['max_rel_errors'] = np.max(rel_errors, axis=0)
         result['max_rel_error_mus'] = test_mus[np.argmax(rel_errors, axis=0)]
-        for name, norm, norm_mu, error, error_mu in zip(error_norm_names,
-                                                        max_norms, max_norm_mus,
-                                                        max_errors[:, -1], max_error_mus[:, -1]):
+        for name, norm, _norm_mu, error, error_mu in zip(error_norm_names,
+                                                         max_norms, max_norm_mus,
+                                                         max_errors[:, -1], max_error_mus[:, -1]):
```
Comment (sdrave) [inline src/pymor/algorithms/error.py:211]: Disagree.

File: src/pymor/algorithms/genericsolvers.py [hunk 5]
```diff
@@ -230,7 +235,7 @@ def lgmres(A, b, x0=None, tol=1e-5, maxiter=1000, M=None, callback=None,
     if b_norm == 0:
         b_norm = 1
 
-    for k_outer in range(maxiter):
+    for k_outer in range(maxiter):  # noqa: B007
```
Comment (sdrave) [inline src/pymor/algorithms/genericsolvers.py:238]: What's the problem here?

File: src/pymor/algorithms/genericsolvers.py [hunk 6]
```diff
@@ -286,7 +291,7 @@ def lgmres(A, b, x0=None, tol=1e-5, maxiter=1000, M=None, callback=None,
             #    solves a minimization problem in the smaller subspace
             #    spanned by W (range) and V (image).
             #
-            #    XXX: Below, I'm lazy and use `lstsq` to solve the
+            #    TODO: Below, I'm lazy and use `lstsq` to solve the
```
Comment (sdrave) [inline src/pymor/algorithms/genericsolvers.py:294]: Disagree. Code is not our's originally, and we do not plan to do anything.

File: src/pymor/algorithms/genericsolvers.py [hunk 7]
```diff
@@ -835,7 +839,7 @@ def lsmr(A, b, damp=0.0, atol=1e-6, btol=1e-6, conlim=1e8,
         str1 = f'{itn:6g} {x.dofs([0])[0]:12.5e}'
         str2 = f' {normr:10.3e} {normar:10.3e}'
         str3 = f'  {test1:8.1e} {test2:8.1e}'
-        print(''.join([str1, str2, str3]))
+        print(f'{str1}{str2}{str3}')
```
Comment (sdrave) [inline src/pymor/algorithms/genericsolvers.py:842]: Disagree.

File: src/pymor/algorithms/genericsolvers.py [hunk 8]
```diff
@@ -967,22 +971,21 @@ def lsmr(A, b, damp=0.0, atol=1e-6, btol=1e-6, conlim=1e8,
...
-                pcount = pcount + 1
-                str1 = f'{itn:6g} {x.dofs([0])[0]:12.5e}'
-                str2 = f' {normr:10.3e} {normar:10.3e}'
-                str3 = f'  {test1:8.1e} {test2:8.1e}'
-                str4 = f' {normA:8.1e} {condA:8.1e}'
-                print(''.join([str1, str2, str3, str4]))
+        if show and ((n <= 40) or (itn <= 10) or (itn >= maxiter - 10)
+                     or (itn % 10 == 0) or (test3 <= 1.1 * ctol)
+                     or (test2 <= 1.1 * atol) or (test1 <= 1.1 * rtol)
+                     or (istop != 0)):
```
Comment (sdrave) [inline src/pymor/algorithms/genericsolvers.py:977]: Disagree.

File: src/pymor/algorithms/greedy.py [hunk 9]
```diff
@@ -134,7 +134,6 @@ def evaluate(self, mus, return_all_values=False):
         If `return_all_values` is `False`, the maximum estimated error as first
         return value and the corresponding parameter as second return value.
         """
-        pass
```
Comment (sdrave) [inline src/pymor/algorithms/greedy.py:137]: Disagree.

File: src/pymor/algorithms/projection.py [hunk 10]
```diff
@@ -126,25 +126,25 @@ def action_apply_basis(self, op):
...
-            if range_basis is None:
-                V = op.apply(source_basis)
-                if isinstance(op.range, NumpyVectorSpace):
-                    from pymor.operators.numpy import NumpyMatrixOperator
-                    return NumpyMatrixOperator(V.to_numpy().T, range_id=op.range.id, name=op.name)
-                else:
-                    from pymor.operators.constructions import VectorArrayOperator
-                    return VectorArrayOperator(V, adjoint=False, name=op.name)
-            else:
+        elif range_basis is None:
```
Comment (sdrave) [inline src/pymor/algorithms/projection.py:137]: Disagree.

**Review by sdrave — CHANGES_REQUESTED** [review]
Hi @pmli, I think there is way to much stuff in this PR that is controversial. I have only marked a few things I, personally, disliked. Some general comments:

- I'd perfer to only enforce rules for which there is an lsp plugin which checks them. I want to be told by my editor that something is wrong. In particular, I don't know about the situation for non-Python files.
- Most importantly, for Python files I very much dislike that, even though we are saying that ruff is our authority regarding code style, making ruff pass is not enough for making CI pass. In particular, since the changes made by autopep8 are maybe not that obvious. I'd rather go back to flake8 than cope with that.
- I believe we should only enforce rules that have a clear benefit for readability/code quality and are easy to follow by a human.

Moving forward, I'd suggest to break this into serveral PRs which are easier to digest/discuss (e.g., one for the mailmap stuff, yaml/toml, ruff). As I said before, much this is a matter of taste. In the end, we should probably put these things up for a vote.

File: src/pymortests/strategies.py [hunk 11]
```diff
@@ -464,11 +464,8 @@ def base_vector_arrays(draw, count=1, dtype=None, max_dim=100):
     length = space.dim
 
     # this lets hypothesis control np's random state too
-    random = draw(hyst.random_module())
-    # scipy performs this check although technically numpy accepts a different range
-    assume(0 <= random.seed < 2**32 - 1)
-    random_correlation = random_correlation_gen(random.seed)
-    rng = np.random.default_rng(0)
+    random = draw(hyst.randoms())
+    rng = np.random.default_rng(random.randint(0, 2**32-1))
```
Comment (sdrave) [inline src/pymortests/strategies.py:468]: Can you say what's going on here?

**Comment by pmli** [comment]
> * [pre-commit] add end-of-file-fixer: I don't see why this is useful.

It's to have a consistent number of empty lines at the end of the files.

> * [ruff] add flake8-bugbear: Generally helpful, but I don't get the change in the `re.sub` call. If this is about the number of arguments: I'm against enforcing the use of keyword arguments.

No, it's specifically about [`re.sub`](https://docs.astral.sh/ruff/rules/re-sub-positional-args/).

> * [ruff] add flake8-comprehensions: It's good that it catches the `all([ .. ])` cases, but I do not want to be restricted how I construct my dictionaries. Using `dict(...)` is an idiom often used to avoid having to write too many quotes.

I agree, I also prefer `dict(foo=foo)` over `{'foo': foo}`. But I think there is no specific rule for `dict` (see [here](https://docs.astral.sh/ruff/rules/unnecessary-collection-call/)).

> * [ruff] add flake8-implicit-str-concat: I think it's perfectly fine to concatenate strings via `+`.

I guess you mean the changes in [`output_error_estimation_with_dwr.py`](https://github.com/pymor/pymor/pull/2070/files#diff-5bd68ff212a613dcb2a2183dd83e9d3713b68cfa5da3bf0ed8e0bf6b26ef3972) and [`trust_region.py`](https://github.com/pymor/pymor/pull/2070/files#diff-6713a2cc51d744f59c5db22d0f4fd90162ef5744673c5de8fb6e69c989faba92) demos? Now that I look at it again, I guess it was enough to remove `+` without adding parentheses (that fix was not automatic).

> * [ruff] add flake8-pytest-style: I see the advantages, but I'm not sure if this isn't too restrictive. Makes the assert blocks very verbose.

I like that very much for debugging (when it's not clear which of the conditions failed in `assert cond1 and cond2 and cond3`).

> * [ruff] add flake8-simplify: Why on earth is `0 < codim` not ok?

[Yoda conditions](https://docs.astral.sh/ruff/rules/yoda-conditions/) are less readable.

**Comment by sdrave** [comment]
> > * [pre-commit] add end-of-file-fixer: I don't see why this is useful.
> 
> It's to have a consistent number of empty lines at the end of the files.

I understand that. But what problem does it solve? Does it reduce bugs or readability?

> > * [ruff] add flake8-bugbear: Generally helpful, but I don't get the change in the `re.sub` call. If this is about the number of arguments: I'm against enforcing the use of keyword arguments.
> 
> No, it's specifically about [`re.sub`](https://docs.astral.sh/ruff/rules/re-sub-positional-args/).

Ok, I don't get why there needs to be a specific rule for `re.sub`, but I guess it's not really used in pyMOR anyway, so be it.
 
> > * [ruff] add flake8-comprehensions: It's good that it catches the `all([ .. ])` cases, but I do not want to be restricted how I construct my dictionaries. Using `dict(...)` is an idiom often used to avoid having to write too many quotes.
> 
> I agree, I also prefer `dict(foo=foo)` over `{'foo': foo}`. But I think there is no specific rule for `dict` (see [here](https://docs.astral.sh/ruff/rules/unnecessary-collection-call/)).

If you look at 1db625230a5e8329be38f741b23319b5a6096e75, you see several changes of this type.

> 
> > * [ruff] add flake8-implicit-str-concat: I think it's perfectly fine to concatenate strings via `+`.
> 
> I guess you mean the changes in [`output_error_estimation_with_dwr.py`](https://github.com/pymor/pymor/pull/2070/files#diff-5bd68ff212a613dcb2a2183dd83e9d3713b68cfa5da3bf0ed8e0bf6b26ef3972) and [`trust_region.py`](https://github.com/pymor/pymor/pull/2070/files#diff-6713a2cc51d744f59c5db22d0f4fd90162ef5744673c5de8fb6e69c989faba92) demos? Now that I look at it again, I guess it was enough to remove `+` without adding parentheses (that fix was not automatic).

For instance, I find the change in src/pymor/bindings/fenics.py, line 616 too strict. Most people don't even know that you can concatenate strings like that.

> 
> > * [ruff] add flake8-pytest-style: I see the advantages, but I'm not sure if this isn't too restrictive. Makes the assert blocks very verbose.
> 
> I like that very much for debugging (when it's not clear which of the conditions failed in `assert cond1 and cond2 and cond3`).

Yes, I know, and I would be willing to enforce this due to this argument. On the other hand, in sylvester.py, this replaces two lines by six three times. This can also harm readability. Especially, when assertions are made in the middle of an algorithm. I feel that this is a significant change that should be discussed in a larger group. Maybe you can put it in a separate PR?

> 
> > * [ruff] add flake8-simplify: Why on earth is `0 < codim` not ok?
> 
> [Yoda conditions](https://docs.astral.sh/ruff/rules/yoda-conditions/) are less readable.

The example given there is ok. But I find `0 < codim` actually much easier to read. I would remove that rule.
Also, I didn't see this yesterday, but I also dislike how the use of dicts is enforced in [referenceelements.py](https://github.com/pymor/pymor/pull/2070/commits/de2ab3749d5075603c539be524f4253710e797a1#diff-f0aa5a9188b23de6034769ec45888944697e1b2b00ad77637e5fb9a5cbb26e14).

File: src/pymortests/strategies.py (hunk 11 above)
Comment (sdrave) [reply src/pymortests/strategies.py:468]: If we both don't understand the code, mybe we should not touch it unless we better understand it. Maybe open a separate issue?

Comment (pmli) [reply src/pymortests/strategies.py:468]: I also don't understand the current code in `main`, so both are equally bad for me in that sense...
What I noticed is that, after removing `np.random.seed(42)` in [`reset_rng`](https://github.com/pymor/pymor/blob/2360b7ca36d5333498cbcc741b300bbe6a8621d6/conftest.py#L61-L64), some tests started failing on CI when they worked fine locally, and after replacing `hyst.random_module` with `hyst.randoms` both CI and running locally worked fine.

**Comment by pmli** [comment]
> > > * [pre-commit] add end-of-file-fixer: I don't see why this is useful.
> > 
> > 
> > It's to have a consistent number of empty lines at the end of the files.
> 
> I understand that. But what problem does it solve? Does it reduce bugs or does it improve readability?

It solves the problem of inconsistency... Also, git complains a bit in the diff when there is no new line at the end.

> > > * [ruff] add flake8-bugbear: Generally helpful, but I don't get the change in the `re.sub` call. If this is about the number of arguments: I'm against enforcing the use of keyword arguments.
> > 
> > 
> > No, it's specifically about [`re.sub`](https://docs.astral.sh/ruff/rules/re-sub-positional-args/).
> 
> Ok, I don't get why there needs to be a specific rule for `re.sub`, but I guess it's not really used in pyMOR anyway, so be it.

There is an explanation in the above link.

> > > * [ruff] add flake8-comprehensions: It's good that it catches the `all([ .. ])` cases, but I do not want to be restricted how I construct my dictionaries. Using `dict(...)` is an idiom often used to avoid having to write too many quotes.
> > 
> > 
> > I agree, I also prefer `dict(foo=foo)` over `{'foo': foo}`. But I think there is no specific rule for `dict` (see [here](https://docs.astral.sh/ruff/rules/unnecessary-collection-call/)).
> 
> If you look at [1db6252](https://github.com/pymor/pymor/commit/1db625230a5e8329be38f741b23319b5a6096e75), you see several changes of this type.

The rule is ignored now. There are actually no instances of `dict()`, `list()`, or `tuple()`, so the rule only complains about `dict(foo=foo)` actually.

> > > * [ruff] add flake8-implicit-str-concat: I think it's perfectly fine to concatenate strings via `+`.
> > 
> > 
> > I guess you mean the changes in [`output_error_estimation_with_dwr.py`](https://github.com/pymor/pymor/pull/2070/files#diff-5bd68ff212a613dcb2a2183dd83e9d3713b68cfa5da3bf0ed8e0bf6b26ef3972) and [`trust_region.py`](https://github.com/pymor/pymor/pull/2070/files#diff-6713a2cc51d744f59c5db22d0f4fd90162ef5744673c5de8fb6e69c989faba92) demos? Now that I look at it again, I guess it was enough to remove `+` without adding parentheses (that fix was not automatic).
> 
> For instance, I find the change in src/pymor/bindings/fenics.py, line 616 too strict. Most people don't even know that you can concatenate strings like that.

And once people learn, they are happy that they don't have to type extra characters, right?

> > > * [ruff] add flake8-pytest-style: I see the advantages, but I'm not sure if this isn't too restrictive. Makes the assert blocks very verbose.
> > 
> > 
> > I like that very much for debugging (when it's not clear which of the conditions failed in `assert cond1 and cond2 and cond3`).
> 
> Yes, I know, and I would be willing to enforce this due to this argument. On the other hand, in sylvester.py, this replaces two lines by six three times. This can also harm readability. Especially, when assertions are made in the middle of an algorithm. I feel that this is a significant change that should be discussed in a larger group. Maybe you can put it in a separate PR?

Ok, it's in #2226 now.

> > > * [ruff] add flake8-simplify: Why on earth is `0 < codim` not ok?
> > 
> > 
> > [Yoda conditions](https://docs.astral.sh/ruff/rules/yoda-conditions/) are less readable.
> 
> The example given there is ok. But I find `0 < codim` actually much easier to read. I would remove that rule. Also, I didn't see this yesterday, but I also dislike how the use of dicts is enforced in [referenceelements.py](https://github.com/pymor/pymor/pull/2070/commits/de2ab3749d5075603c539be524f4253710e797a1#diff-f0aa5a9188b23de6034769ec45888944697e1b2b00ad77637e5fb9a5cbb26e14).

Ok, the rule is ignored now.

File: src/pymortests/strategies.py (hunk 11 above)
Comment (sdrave) [reply src/pymortests/strategies.py:468]: That is strange. I'd suggest to also move this into a different issue/PR to be able to look at this more carefully. (And also leave the `np.random.seed(42)` in for this PR.)

**Comment by sdrave** [comment]
> > > > * [pre-commit] add end-of-file-fixer: I don't see why this is useful.
> > > 
> > > 
> > > It's to have a consistent number of empty lines at the end of the files.
> > 
> > 
> > I understand that. But what problem does it solve? Does it reduce bugs or does it improve readability?
> 
> It solves the problem of inconsistency... Also, git complains a bit in the diff when there is no new line at the end.

IMHO, consistency is not a value in itself. It's good if it improves readability or helps prevent errors. I don't see this here. But if it's important to you, we can leave it in.

> > Ok, I don't get why there needs to be a specific rule for `re.sub`, but I guess it's not really used in pyMOR anyway, so be it.
> 
> There is an explanation in the above link.

Ok, I see.

> > > > * [ruff] add flake8-implicit-str-concat: I think it's perfectly fine to concatenate strings via `+`.
> > For instance, I find the change in src/pymor/bindings/fenics.py, line 616 too strict. Most people don't even know that you can concatenate strings like that.
> 
> And once people learn, they are happy that they don't have to type extra characters, right?

Maybe. Maybe they are unhappy because of all these strict rules one has to adhere to to write pyMOR code. But maybe it improves readability, so we can leave it in.

---

One last thing I have been forgetting about: I'm not sure I like

[ruff] add Ruff-specific rules 

Sure, the use of iterable unpacking is quite elegant. I guess it's also more performant (but not in a relevant way). But it really is a quite advanced idiom that probably few people in our community know. So for them, it actually decreases readability. So maybe it is better not enforce it.

**Comment by drittelhacker** [comment]
> > > > > * [pre-commit] add end-of-file-fixer: I don't see why this is useful.
> > > > 
> > > > 
> > > > It's to have a consistent number of empty lines at the end of the files.
> > > 
> > > 
> > > I understand that. But what problem does it solve? Does it reduce bugs or does it improve readability?
> > 
> > 
> > It solves the problem of inconsistency... Also, git complains a bit in the diff when there is no new line at the end.
> 
> IMHO, consistency is not a value in itself. It's good if it improves readability or helps prevent errors. I don't see this here. But if it's important to you, we can leave it in.
While you are completely right about consistency in general, in cases like this, it can help to keep a cleaner git history and make pull requests easier to read/review. You just don't need to filter out meaningless whitespace changes.


> > > > > * [ruff] add flake8-implicit-str-concat: I think it's perfectly fine to concatenate strings via `+`.
> > > > >   For instance, I find the change in src/pymor/bindings/fenics.py, line 616 too strict. Most people don't even know that you can concatenate strings like that.
> > 
> > 
> > And once people learn, they are happy that they don't have to type extra characters, right?
> 
> Maybe. Maybe they are unhappy because of all these strict rules one has to adhere to to write pyMOR code. But maybe it improves readability, so we can leave it in.
I agree that too many rules are scaring people off. However, these tools can also help. If linters already tell  me how to adhere to the rules before I have to enter lengthy discussions with the main devs why my code is bad, once my PR is under revision, this could actually help me contribute.

**Comment by sdrave** [comment]
> I agree that too many rules are scaring people off. However, these tools can also help. If linters already tell me how to adhere to the rules before I have to enter lengthy discussions with the main devs why my code is bad, once my PR is under revision, this could actually help me contribute.

I never argued against using linters for Python code. And if it's a rule that is needed to be followed for the code to be accepted into pyMOR, then it's of course desirable to have a linter rule for that. In each case, I'm never arguing whether or not to have a certain rule for the linter, but whether or not a certain rule needs to be followed for the code to be accepted into pyMOR.

**Pull request merged**